### DISCUSSION TITLE
Is there a way to replace emscripten stub functions?

### DISCUSSION THREAD

**Discussion by iliakan** [emscripten-core/emscripten#18728] [Q&A] [answered]
I'm trying to port a program that calls `fork` and `execve` on some occasions.

I'd like to show my error and stop the execution in cases of such calls.

Here's the stub for e.g. `fork`:

https://github.com/emscripten-core/emscripten/blob/main/system/lib/libc/emscripten_libc_stubs.c#L165

I'd like the build to use my `fork` instead. 

That is, replace this symbol build-time OR after the build (or in any other way).
How ? ;)

**Comment by sbc100** [comment, marked as answer]
If you include your own definition of `fork` in your program (i.e. in one of your object files, or libraries) the linker will choose your definition over the stub one. 

Alternatively if you set `-sALLOW_UNIMPLEMENTED_SYSCALLS=0` the stub library will not be included at all, and you will be forced to implement any symbols your program depends on.

From settings.js:

```                                                                       
// Include unimplemented JS syscalls to be included in the final output.  This   
// allows programs that depend on these syscalls at runtime to be compiled, even 
// though these syscalls will fail (or do nothing) at runtime.                      
var ALLOW_UNIMPLEMENTED_SYSCALLS = true;  
```

**Reply by iliakan** [reply]
Does this flag remove only emscripten/system/lib/libc/emscripten_syscall_stubs.c or something else too?

Here's the line, but I'm not sure which C files it covers:

https://github.com/emscripten-core/emscripten/blob/9fdc94ad3e3c89558fd251048e8ae2c2ca408dc1/tools/system_libs.py#L2111-L2112

**Reply by sbc100** [reply]
Yup, that library includes that file: https://github.com/emscripten-core/emscripten/blob/9fdc94ad3e3c89558fd251048e8ae2c2ca408dc1/tools/system_libs.py#L1938-L1941

**Reply by iliakan** [reply]
And then I should write my own C-files implementing those functions? All of them?

**Reply by sbc100** [reply]
Basically it removes all the implements of things that emscripten doesn't implement.  Anything where we just return an error instead of doing something useful.

We include these stubs by default to make more programs link out-of-the-box, but for folks that want to be precise about what parts of their program are actually needed it can be better to set `ALLOW_UNIMPLEMENTED_SYSCALLS` and then remove any references to undefined symbols.   For example, if a program calls `fork` it might be preferable to `ifdef` out all of code surrounding that call, rather then just have fork fail at runtime, but we we want to give users both options.

**Reply by sbc100** [reply]
If you just want to implement one of them you can probably just define that function and it should work.    I don't think you should need to use `ALLOW_UNIMPLEMENTED_SYSCALLS`.

**Reply by sbc100** [reply]
Why happens if you just define `fork` in your code?  Does it not work?

**Reply by iliakan** [reply]
Here's the error if I do that without any flags:
```
wasm-ld: error: duplicate symbol: fork
>>> defined in my.a(run-command.o)
>>> defined in /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libstubs-debug.a(emscripten_libc_stubs.o)
```

**Reply by sbc100** [reply]
Ah, this is because your program also uses some other function in the `emscripten_libc_stub.s.c` file.  I think the simplest solution here would be to mark the stub function(s) as weak.  

I think musl provides a `weak` macro which expands to `__attribute__((__weak__))` to I think in this file you can just prefix the declaration of `fork` with the word `weak`:  `weak pid_t fork(void) {`.

Whenever you make any changes to the source code for any of the system libraries you also need to make sure they get rebuilt.   The simplest way to do that would be `./emcc --clear-cache` which removes all libraries.   But you can be more precise with `./embuilder build <name> --force` where `name` is one of the library names, which you can find from `./embuilder --help`.

**Reply by iliakan** [reply]
You mean I should change the file `emscripten_libc_stub.s.c`?

**Reply by sbc100** [reply]
The stubs are defined in `system/lib/libc/emscripten_libc_stubs.c`

**Reply by iliakan** [reply]
I'd just prefer not to patch emscripten itself for the purposes of a single project ;) 

Or maybe we should update Emscripten to make all these functions weak? 
To allow redefinition for other projects?

**Comment by iliakan** [comment]
Right now I'm building in two steps:
1. `emmake make all ... CFLAGS="$LINK_AND_COMPILE_CFLAGS"`
2. `emcc ${=LINK_AND_COMPILE_CFLAGS} ${=COMPILE_CFLAGS} *.o -o my.js`

Where should I put the `-sALLOW_UNIMPLEMENTED_SYSCALLS=0`, if I am to try it? ;)

Should it go to (1) or (2)? In CFLAGS or LDFLAGS?

P.S. I tried a couple of things, didn't work, so would like to ensure I do that right.

**Reply by sbc100** [reply]
`ALLOW_UNIMPLEMENTED_SYSCALLS` is a link time setting, so it would go in `LDFLAGS`.  We even have a nice warning if you try to put it in `CFLAGS`:

```
$ ./emcc -sALLOW_UNIMPLEMENTED_SYSCALLS=0 test/hello_world.c -c
emcc: warning: linker setting ignored during compilation: 'ALLOW_UNIMPLEMENTED_SYSCALLS' [-Wunused-command-line-argument]
```

(But I think the marking of the stub functions as weak is the likely the way you want to go).

**Reply by iliakan** [reply]
Ok, I actually tried that. Got a bunch or errors for other symbols ;)

```
wasm-ld: error: daemon.o: undefined symbol: getpwnam
wasm-ld: error: daemon.o: undefined symbol: getgrnam
wasm-ld: error: libgit.a(ident.o): undefined symbol: getpwuid
wasm-ld: error: daemon.o: undefined symbol: getpwnam
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(getpgid.o): undefined symbol: __syscall_getpgid
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(getrlimit.o): undefined symbol: __syscall_prlimit64
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(getrlimit.o): undefined symbol: __syscall_ugetrlimit
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(getuid.o): undefined symbol: __syscall_getuid32
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(link.o): undefined symbol: __syscall_link
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(setsid.o): undefined symbol: __syscall_setsid
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(uname.o): undefined symbol: __syscall_uname
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-debug.a(waitpid.o): undefined symbol: __syscall_wait4
wasm-ld: error: /opt/projects/git/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libsockets.a(setsockopt.o): undefined symbol: __syscall_setsockopt
```

So I am to implement all that stuff, right?

**Reply by sbc100** [reply]
I think the best approach, if you just want to impement/modify, `fork` is to mark is as `weak` and not to take the `ALLOW_UNIMPLEMENTED_SYSCALLS` appoach.

As you can see, `ALLOW_UNIMPLEMENTED_SYSCALLS` opts out of the emscrpiten stubs, which would mean you would need to fill all of them in again somehow, or modify your program not to depend on them.

**Reply by iliakan** [reply]
Why not label all functions as weak in stubs, in upstream Emscripten?

Looks like a good way to go. Stubs that can easily be replaced without patching :yammy!:

**Reply by sbc100** [reply]
Yup, that makes sense me.

**Comment by sbc100** [comment]
BTW, if you are going to try to implement `fork` for in emscripten you might have hard time, depending on what your program is doing.  If you program is just using `fork` immediately followed by `exec` you might want to look into some perhaps easier solutions such as `vfork` or (even better) `posix_spawn`.   These don't require the complete cloning the current process.

**Reply by iliakan** [reply]
Right now the program dies in a strange manner when it tries to use `fork` and few other not implemented syscalls.

Some stuff I rewrote using WASM module instantiation to allow nested execution.

For other, less important functionality that relies on `fork`, I'd like to make `fork` show a nice error message, explaining that the functionality is limited, because it's in the browser environment. And then kill the process, to avoid other odd side-effects ;)

That's all I plan to put into a new, rewritten fork ;)

**Reply by iliakan** [reply]
The `emcc --clear-cache` worked, thanks (in correct emsdk env).

**Comment by iliakan** [comment]
So I added `__attribute__((__weak__))` to `pid_t fork(void)` in `emscripten_libc_stubs.c`, and it compiled!

My `fork` is working now.

Does it seem to be a good PR if I add this  `__attribute__((__weak__))` to all other functions in this file?

**Reply by sbc100** [reply]
Yes please, that would great.   Perhaps we can use the musl macro `weak` instead though?

**Reply by iliakan** [reply]
Do we have to?

Actually, I tried `include <features.h>` and then just `weak`, but it didn't fly :/

**Reply by iliakan** [reply]
```c
#include <features.h>
...
weak pid_t fork(void) {
```

 unknown type name 'weak'

**Reply by sbc100** [reply]
Fair enough, looks like our stub library doesn't inherit the musl internal header path.  The spelled out attribute sounds fine to me, or you could add #define weak ...` at the top of the file.

**Reply by iliakan** [reply]
Suggested as https://github.com/emscripten-core/emscripten/pull/18736
